launcher/agent/agent.go

```diff
@@ -152,10 +210,62 @@ func (a *agent) Attest(ctx context.Context, opts AttestAgentOpts) ([]byte, error
 	return resp.ClaimsToken, nil
 }
 
-func (a *agent) attest(nonce []byte, cel []byte) (*pb.Attestation, error) {
-	a.tpmMu.Lock()
-	defer a.tpmMu.Unlock()
-	return a.fetchedAK.Attest(client.AttestOpts{Nonce: nonce, CanonicalEventLog: cel, CertChainFetcher: http.DefaultClient})
+type tpmAttestRoot struct {
+	tpmMu     sync.Mutex
+	fetchedAK *client.Key
+	tpm       io.ReadWriteCloser
+}
+
+func (t *tpmAttestRoot) Extend(c cel.Content, l *cel.CEL) error {
+	return l.AppendEventPCR(t.tpm, cel.CosEventPCR, defaultCELHashAlgo, c)
+}
+
+func (t *tpmAttestRoot) Attest(nonce []byte) (any, error) {
+	t.tpmMu.Lock()
+	defer t.tpmMu.Unlock()
+
+	return t.fetchedAK.Attest(client.AttestOpts{
+		Nonce:            nonce,
+		CertChainFetcher: http.DefaultClient,
+	})
+}
+
+type tdxAttestRoot struct {
+	tdxMu     sync.Mutex
+	qp        *tg.LinuxConfigFsQuoteProvider
+	tsmClient configfsi.Client
+}
+
+func (t *tdxAttestRoot) Extend(c cel.Content, l *cel.CEL) error {
+	return l.AppendEventRTMR(t.tsmClient, cel.CosRTMR, c)
+}
+
+func (t *tdxAttestRoot) Attest(nonce []byte) (any, error) {
+	t.tdxMu.Lock()
+	defer t.tdxMu.Unlock()
+
+	var tdxNonce [tlabi.TdReportDataSize]byte
+	copy(tdxNonce[:], nonce)
+
+	rawQuote, err := tg.GetRawQuote(t.qp, tdxNonce)
+	if err != nil {
+		return nil, err
+	}
+
+	ccelData, err := os.ReadFile("/sys/firmware/acpi/tables/data/CCEL")
+	if err != nil {
+		return nil, err
+	}
+	ccelTable, err := os.ReadFile("/sys/firmware/acpi/tables/CCEL")
```

Comment:
I am surprised there's no utilities for this in go-tdx-guest. There should be.

Reply:
noted